### ISSUE TITLE
[css] provide css formatter

### ISSUE THREAD

**Issue by isidorn** [microsoft/vscode#19166]
Refs #19047

Our css does not support formating a range, thus `formatOnPaste` is useless for CSS.
And this is actually the only use case I personally have for `formatOnPaste`, since I always copy rules from chrome dev tools and they are always wrongly indented. However `formatOnPaste` does not help me here

**Comment by jrieken** [comment]
Life is unfair. There is also `formatOnType` for `json` and no party without hangover

**Comment by aeschli** [comment]
There are extensions that provide formatting css. I'd suggest you try one of these.
IMO we should not package it all to VSCode. We want to share the maintenance work and want to keep the VSCode download small.

**Comment by isidorn** [comment]
Sure but then we should somehow communicate this to our users of `formatOnPaste` setting

**Comment by aeschli** [comment]
@isidorn Can you create a separate issue for the suggestion to communicate it better (or rename this issue)

**Comment by isidorn** [comment]
@aeschli part of that is already captured in https://github.com/Microsoft/vscode/issues/19160

Apart from that I would expect that our docs cover this - created this one https://github.com/Microsoft/vscode-docs/issues/809

**Comment by vkWeb** [comment]
CSS formatting is an integral part of a text editor. I request you @jrieken that you'll look into this.

**Comment by aeschli** [comment]
@vkWeb CSS formatting is available as an extension. Check out the [markerplace](https://marketplace.visualstudio.com/search?term=css%20formatter&target=vscode&category=All%20categories&sortBy=Relevance)

**Comment by vkWeb** [comment]
@aeschli Ya I know that :). I was trying to say that a code editor must have some kind of default formatting options for CSS. VSCode lacks that. Though, It's not a big deal ;)

**Comment by levinit** [comment]
need css formatter:pray:

**Comment by vkWeb** [comment]
@levinit You can use [Prettier](https://prettier.io/) - A great code formatter for JS, CSS & Typescript. 
I'm closing this issue @aeschli 👍

**Comment by levinit** [comment]
@vkWeb thx ,I have installed it, it works.(although I hope vscode integrates css formatter:yum:)

**Comment by vkWeb** [comment]
@levinit Once you get into general purpose programming or back-end stuff. You'll realize VSCode is a gem. A true gem!  ❤️  😃

**Comment by Neutrino-Sunset** [comment]
Is there actually a css formatter in the marketplace that obeys the editorconfig indentation settings? Because I've tried about half a dozen and haven't found one that does yet.

**Comment by stevensacks** [comment]
The extensions that "handle formatting css" do not work *while you type*, and even then none of them follow the rules set in your prettier, scss-lint, editorconfig, etc.

Your Javascript editor formats as you type. CSS should do the same.

And just to drive the point home, here is VSCode on the left and WebStorm on the right. Which would you rather use? But please, 

![code-vs-webstorm](https://user-images.githubusercontent.com/162252/43580109-34cd8928-968f-11e8-8181-67fbe45e72ea.gif)

Unfortunately, prettier is not working for scss/css files in WebStorm, as you can see. I have prettier installed along with scss-lint, and Beautify css/sass/scss - none of them do anything.

If I use the Format Code key command it formats css in a way that breaks all the scss-lint rules I have (it removes all line breaks from the entire file).

Fact of the matter is, WebStorm is eating your lunch on a feature that should be native to the application which is BASIC formatting of CSS as shown in the above gif. Don't pass the buck just because CSS-in-JS is so popular.

**Comment by aeschli** [comment]
BTW there's only so much the VSCode team can work on and maintain. We prioritize on the number of votes on the issues (👍 on the issue description). That's also why we keep these bugs open.

Besides, not everything should and can be bundled with VSCode. That's why we have an extension API and a marketplace. That lets us share the maintenance work, encourage participation & innovation and want to keep the VSCode download small.

Please support these extension authors by using their extensions and give feedback. File issues describing the functionality they are missing (e.g. don't remove my empty lines).

Adding 'format on save' requires implementing a [OnTypeFormattingEditProvider](https://github.com/Microsoft/vscode/blob/c09911add3efda281261aa710b2b53fad32f1144/src/vs/vscode.d.ts#L2984)

**Comment by kanlukasz** [comment]
I understand that it's not possible to bundle everything in one BUT if you have built-in HTML and JS formatter and you don't have CSS formatter, something is really illogical here. JS, HTML, CSS - it's all very closely related.

You say - use add-ons from Marketplace.
I'd really like to do that, but please see how some add-ons work.
Many of them do not have proper documentation and many bugs. Many things don't work properly.
Add-ons developers leave their projects. I often see an decription like: _"I don't have time to work on it, if you want you can take over this project"_
Maybe I'm too pessimistic but thats how it works, unfortunately...

To sum up - the built-in CSS formatter is something really important. I hope you add it

**Comment by aeschli** [comment]
Closing as we still have no plans to include a CSS formatter as discussed in https://github.com/microsoft/vscode/issues/19166#issuecomment-409928738

**Comment by ArturoDent** [comment]
Nit:  In the description for `CSS > Format: Newline Between Selectors`, it says

> Separate selectors with a new line.

I believe that should say `newline` rather than `new line`.

@isidorn
